src/gtk3-porting-guide.md

```diff
@@ -0,0 +1,780 @@
+
+    from gi.repository import Pango, PangoCairo
+
+    cairo_context = ...
+    pango_layout = PangoCairo.create_layout(cairo_context)
+    fd = Pango.FontDescription('Sans')
+    fd.set_size(12 * Pango.SCALE)
+    pango_layout.set_font_description(fd)
+    pango_layout.set_text('Hello World', -1)
+    cairo_context.set_source_rgb(1, 0, 0)
+    PangoCairo.update_layout(cairo_context, pango_layout)
+    PangoCairo.show_layout(cairo_context, pango_layout)
+
+The get\_extents() method if different in PangoCairo. It calculates an
+extent as a Rectangle, but doesn't return anything. There is a method,
+get\_logical\_extents() that returns a Rectangle. Alas, it is not
+necessarily available after v1.16. Note that Rectangle is not a list but
+a class with methods for get\_x(), get\_y(), get\_width(), and
+get\_height(), so you cannot iter over it.
+
+Note that the <samp>cairo.Region</samp> will no longer work in GTK+3
+
+#### Replacing pixmaps with Cairo
+
+You need to replace your pixmaps with Cairo in GTK+ 3. For an example on
+how this is done, see:
+<http://developer.gnome.org/gtk3/3.5/ch24s02.html#idp129615008>
+
+### Taking a screenshot and making a thumbnail
+
+To make a screenshot of the window:
+
+    width, height = window.get_width(), window.get_height()
+    thumb_surface = Gdk.Window.create_similar_surface(window,
+                                                      cairo.CONTENT_COLOR,
+                                                      width, height)
+
+    thumb_width, thumb_height = style.zoom(100), style.zoom(80)
+    cairo_context = cairo.Context(thumb_surface)
+    thumb_scale_w = thumb_width * 1.0 / width
+    thumb_scale_h = thumb_height * 1.0 / height
+    cairo_context.scale(thumb_scale_w, thumb_scale_h)
+    Gdk.cairo_set_source_window(cairo_context, window, 0, 0)
+    cairo_context.paint()
+    thumb_surface.write_to_png(png_path_or_filelike_object)
+
+### Creating a video widget
+
+Some necessary changes include:
+
+Using
+
+`get_property('window').get_xid()`
+
+Instead of
+
+`window.xid`
+
+Using
+
+`set_double_buffered(False)`\
+`set_app_paintable(True)`
+
+Instead of
+
+`unset_flags(gtk.DOUBLE_BUFFERED)`\
+`set_flags(gtk.APP_PAINTABLE)`
+
+Hacks to help in porting
+========================
+
+### Use the same keyboard and mouse
+
+If you have an XO, I'm sure you want to take a look at
+[this](User:Humitos/x2x "wikilink")...
+
+### Use Extended Python debugger
+
+**epdb** library is useful to inspect the code while the Activity is
+running.
+
+`sudo yum install python-epdb`
+
+After that I put some trace point in the code where I can stop and make
+my tests by doing this:
+
+`import epdb;epdb.set_trace()`
+
+Finally I run Get Books Activity from the Terminal Activity to be able
+to write some code on a shell. This is the command that I use:
+
+`sugar-launch org.laptop.sugar.GetBooksActivity`
+
+See also [Development
+Team/Debugging](Development_Team/Debugging "wikilink").
```

Comment:
Please fix the link

Reply:
couldn't find the correct link.
removed the line instead.